File: cryptopals/set_1/challenge_4.py
import os
from cryptopals.set_1 import get_most_likely_xor_char


def get_most_likely_line():
    with open(os.path.join(os.path.dirname(__file__), "challenge_4.txt"), "rb") as f:
        max_score = None
        best_line_number = None
        best_line = None
        best_line_char = None
        best_line_output = None
        for i, line in enumerate(f):
            char, score, s = get_most_likely_xor_char(line.rstrip())
            if max_score is None or score > max_score:
                max_score = score
                best_line_number = i
                best_line = line.rstrip()
                best_line_char = char
                best_line_output = s
    return best_line_number, best_line_char, max_score, best_line, best_line_output

File: pkg/set1/challenge_4.go
package set1

import (
	"bufio"
	"log"
	"os"
	"strings"
)

func FileToArray(filePath string) []string {
	file, err := os.Open(filePath)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	arr := make([]string, 0)
	for scanner.Scan() {
		arr = append(arr, strings.TrimSpace(scanner.Text()))
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}
	return arr
}

// MostLikelyXorLine gets the line with the best xor score/character
func MostLikelyXorLine() (int, byte, float64, string) {
	var maxScore float64
	var bestLineNumber int
	var bestLineOutput string
	var bestLineChar byte
	isFirst := true
	for i, line := range loadChallenge4() {
		byteChar, score, xordLine := MostLikelyXorChar(HexToBytes(line))
		if isFirst || score > maxScore {
			maxScore = score
			bestLineNumber = i
			bestLineOutput = string(xordLine)
			bestLineChar = byteChar
			isFirst = false
		}
	}
	return bestLineNumber, bestLineChar, maxScore, bestLineOutput
}
func loadChallenge4() []string {
	return FileToArray("/Users/dboren/dev/go/src/github.com/davidboren/cryptopals/data/set1/challenge_4.txt")
}
